File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys

while True:
    stack = []
    data = sys.stdin.readline().rstrip()
    if data[0] == '.' and len(data) == 1:
        break
    else:
        for i in range(len(data)):
            if data[i] == '(' or data[i] == '[':
                stack.append(data[i])
            else:
                if data[i] == ')':
                    if stack:
                        if stack[-1] == '(':
                            stack.pop()
                        else:
                            sys.stdout.write("no\n")
                            break
                    else:
                        sys.stdout.write("no\n")
                        break

                elif data[i] == ']':
                    if stack:
                        if stack[-1] == '[':
                            stack.pop()
                        else:
                            sys.stdout.write("no\n")
                            break
                    else:
                        sys.stdout.write("no\n")
                        break
        else:
            if stack:
                sys.stdout.write("no\n")
            else:
                sys.stdout.write("yes\n")

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuilder sb = new StringBuilder();

        while (true) {
            String[] input = br.readLine().split("");
            if (input[0].equals(".")) {
                break;
            }

            Stack<String> stack = new Stack<>();
            boolean flag = false;
            for (String data : input) {
                if (data.equals("[") || data.equals("(")) {
                    stack.push(data);
                    continue;
                }

                if (data.equals("]")) {
                    if (stack.isEmpty()) {
                        flag = true;
                        break;
                    }
                    if (stack.peek().equals("[")) {
                        stack.pop();
                        continue;
                    } else {
                        flag = true;
                        break;
                    }
                }

                if (data.equals(")")) {
                    if (stack.isEmpty()) {
                        flag = true;
                        break;
                    }
                    if (stack.peek().equals("(")) {
                        stack.pop();
                    } else {
                        flag = true;
                    }
                }
            }

            if (flag) {
                sb.append("no\n");
                continue;
            }
            if (stack.isEmpty()) {
                sb.append("yes\n");
            } else {
                sb.append("no\n");
            }
        }
        bw.write(sb.toString());
        bw.close();
        br.close();
    }
}
